### PR TITLE
make unaligned_reference a hard error

### LINKED ISSUE #82523: Tracking Issue for future-incompatibility warning `unaligned_references`
This is a tracking issue for the future-incompatibility warning `unaligned_references`.
This warning will fire for code like the following:
```rust
#[repr(packed)]
struct Foo1 {
    bar: u8,
    baz: usize
}

let foo = Foo1 { bar: 1, baz: 2 };
// Direct field accesses are fine.
let val = foo.baz;
// However, creating a reference is not.
let brw = &foo.baz; //~WARN reference to packed field is unaligned
// Format strings implicitly create references.
println!("{}", foo.baz); //~WARN reference to packed field is unaligned
```
The reason this pattern is being phased out is that Rust requires references to always be aligned; creating an unaligned reference falls under the "creating an invalid value" clause in [the Rust definition of Undefined Behavior](https://doc.rust-lang.org/reference/behavior-considered-undefined.html). Fields of packed structs are not necessarily properly aligned. Hence creating a reference to a field of a packed struct can cause UB, even if it is never used, and even inside an `unsafe` block. This is a soundness bug, which is fixed by deprecating and eventually disallowing this pattern.

Previously, a future-incompatibility warning was emitted when creating references to packed fields outside an unsafe block; however, that warning was incorrectly silenced inside unsafe blocks.

To fix this code, it needs to stop creating a reference to a packed field. The alternative is to either just *copy* the packed field by adding curly braces (the compiler knows how to do that despite lack of alignment), or to create a raw pointer:
```rust
let mut foo = Foo1 { bar: 1, baz: 2 };

let brw = std::ptr::addr_of!(foo.baz); // Create an immutable raw pointer to the packed field.
let val = unsafe { brw.read_unaligned() }; // Perform an unaligned read of that pointer.
let brw_mut = std::ptr::addr_of_mut!(foo.baz); // Create a mutable raw pointer to the packed field.
unsafe { brw_mut.write_unaligned(val+1); } // Perform an unaligned write to that pointer.

// For formatting, adding curly braces means that a copy of the field is made, stored
// in a (properly aligned) temporary, and a reference to that temporary is being formatted.
println!("{}", {foo.baz});
// This is equivalent to the more verbose
println!("{}", {let copy = foo.baz; copy});
```

For further background, see https://github.com/rust-lang/rust/issues/46043, #27060.

### PR DESCRIPTION
**Pull request by RalfJung**
The `unaligned_references` lint has been warn-by-default since Rust 1.53 (https://github.com/rust-lang/rust/pull/82525) and deny-by-default with mention in cargo future-incompat reports since Rust 1.62 (https://github.com/rust-lang/rust/pull/95372). Current nightly will become Rust 1.66, so (unless major surprises show up with crater) I think it is time we make this a hard error, and close this old soundness gap in the language.

EDIT: Turns out this will only land for Rust 1.67, so there is another 6 weeks of time here for crates to adjust.

Fixes https://github.com/rust-lang/rust/issues/82523.

### CODE REVIEW

File: compiler/rustc_error_codes/src/error_codes/E0791.md [hunk 1]
```diff
@@ -0,0 +1,43 @@
+An unaligned references to a field of a [packed] struct got created.
+
+[packed]: https://doc.rust-lang.org/reference/type-layout.html#the-alignment-modifiers
+
+### Example
```
Comment (GuillaumeGomez) [inline compiler/rustc_error_codes/src/error_codes/E0791.md:5]: Please follow the format of the other error codes.

Comment (RalfJung) [reply compiler/rustc_error_codes/src/error_codes/E0791.md:5]: FWIW https://github.com/rust-lang/rust/blob/03f6bb63669d099c57808628b226716ab572e22e/compiler/rustc_error_codes/src/error_codes.rs#L8 links to a closed (not merged) PR with a defunct 'rendered' link. I don't know what the format is.

Comment (GuillaumeGomez) [reply compiler/rustc_error_codes/src/error_codes/E0791.md:5]: I didn't even know there was a RFC. Normally the format is enforced through some checks (which isn't great). For example in [this one](https://github.com/rust-lang/rust/blob/master/compiler/rustc_error_codes/src/error_codes/E0069.md):

```
Small sum-up of the error.

Erroneous code example:

[code example]

Explanation of what went wrong.

[fixed code example (if applicable)]

Extra information (like a link to the Rust Book for example)
```

Comment (RalfJung) [reply compiler/rustc_error_codes/src/error_codes/E0791.md:5]: Should this link to https://rust-lang.github.io/rfcs/1567-long-error-codes-explanation-normalization.html instead ?

Comment (GuillaumeGomez) [reply compiler/rustc_error_codes/src/error_codes/E0791.md:5]: Yes that would be a huge improvement!

Comment (RalfJung) [reply compiler/rustc_error_codes/src/error_codes/E0791.md:5]: > I didn't even know there was a RFC.

You are the author of that PR...?

Anyway I will update the link and also the error description. Just waiting for the try build to complete.

Comment (GuillaumeGomez) [reply compiler/rustc_error_codes/src/error_codes/E0791.md:5]: ... Don't trust my memory.

Thanks for sending the fix.

**Comment by RalfJung** [comment]
@craterbot check

File: compiler/rustc_error_codes/src/error_codes/E0791.md (hunk 1 above)
Comment (RalfJung) [reply compiler/rustc_error_codes/src/error_codes/E0791.md]: @GuillaumeGomez please take a look at the new error code page.

File: compiler/rustc_error_codes/src/error_codes/E0791.md [hunk 2]
```diff
@@ -0,0 +1,65 @@
+An unaligned references to a field of a [packed] struct got created.
+
+[packed]: https://doc.rust-lang.org/reference/type-layout.html#the-alignment-modifiers
```
[the remark this answers is not in the document]
Comment (RalfJung) [reply compiler/rustc_error_codes/src/error_codes/E0791.md:3]: Ah fair we already have the other links there.

Thanks for taking a look!

**Comment by RalfJung** [comment]
@craterbot run mode=check-only crates=https://crater-reports.s3.amazonaws.com/pr-102513/retry-regressed-list.txt

**Comment by oli-obk** [comment]
Various versions of the `tendril` crate are causing 380 of the 470 regressions (they actually do use unaligned references). The rest are various smaller (<=5 dependents) crates or just spurious build.rs/proc macro failures

**Comment by RalfJung** [comment]
Looks like this got fixed in tendril 0.4.3 half a year ago by [removing the use of `packed`](https://github.com/servo/tendril/commit/bdd07bfcc8b9bd43961248e92a82ed2b936e1868). So at least for the users of the 0.4 series, there is a semver-compatible fix available. According to download stats 0.4.3 is the only version downloaded in relevant quantities, though the crater run shows that a bunch of stale lockfiles still contain 0.4.1 and older versions.

So this is another case of, crater also tests unmaintained parts of the ecosystem with a lockfile, and there is basically nothing we can do to get those parts to update.

**Comment by RalfJung** [comment]
Here are the tendril version stats from the regressed crate
```
      1 Checking tendril v0.1.6
      7 Checking tendril v0.2.4
     46 Checking tendril v0.3.1
     15 Checking tendril v0.4.0
    294 Checking tendril v0.4.1
```
So a bit more than 50 don't have a semver-compatible update. These are almost all on github; in terms of crates.io there are only 6:
```
reg/snoot/0.1.2/try#c34df645207dd0b6887b2312cf109109ba3a0ffb.txt
reg/webgl_generator/0.2.0/try#c34df645207dd0b6887b2312cf109109ba3a0ffb.txt
reg/femto-feed/0.3.0/try#c34df645207dd0b6887b2312cf109109ba3a0ffb.txt
reg/rbatis_xml_parser/0.1.0/try#c34df645207dd0b6887b2312cf109109ba3a0ffb.txt
reg/aocl/0.1.1/try#c34df645207dd0b6887b2312cf109109ba3a0ffb.txt
reg/html2runes/1.0.1/try#c34df645207dd0b6887b2312cf109109ba3a0ffb.txt
```

**Comment by RalfJung** [comment]
@rust-lang/lang I'd like to get approval for making the 'unaligned_references' lint a hard error. It has been warn-by-default since Rust 1.53 (https://github.com/rust-lang/rust/pull/82525) and deny-by-default with mention in cargo future-incompat reports since Rust 1.62 (https://github.com/rust-lang/rust/pull/95372). Current nightly will become Rust 1.66. This finally closes one of our oldest soundness gaps, an oversight that goes back all the way to Rust 1.0.

See above for a detailed analysis of the crater results, the summary is that the majority of the regressions (380 of 470) is caused by a single crate, `tendril`.  Its latest version 0.4.3 is fixed, and 85% of the `tendril` regressions would be fixed by a simple `cargo update`. The remaining crate regressions are rarely used and/or unmaintained crates.

**Comment by scottmcm** [comment]
@rfcbot fcp merge

I agree; we've been moving this way for a *long* time, and cargo future-incompat was the last thing we were waiting on.

**Comment by scottmcm** [comment]
@rfcbot fcp cancel

**Comment by scottmcm** [comment]
@rfcbot fcp merge

I agree; we've been moving this way for a long time, and cargo future-incompat was the last thing we were waiting on.

**Comment by nikomatsakis** [comment]
@rfcbot fcp reviewed

**Comment by RalfJung** [comment]
FCP will pass soon... @cjgillot could I ask for a review of the code side here? :)

File: compiler/rustc_error_codes/src/error_codes/E0791.md [hunk 3]
```diff
@@ -0,0 +1,64 @@
+An unaligned references to a field of a [packed] struct got created.
```
Comment (oli-obk) [inline compiler/rustc_error_codes/src/error_codes/E0791.md:1]:
```suggestion
An unaligned reference to a field of a [packed] struct got created.
```

**Review by oli-obk — COMMENTED** [review]
Just a typo, rest lgtm

**Comment by cjgillot** [comment]
r=me once FCP completes

**Comment by Dylan-DPC** [comment]
failed in [rollup](https://github.com/rust-lang/rust/pull/103436#issuecomment-1288178773)

@bors r-

**Comment by RalfJung** [comment]
Looks like winapi 0.2.8, which has not received a semver-compatible update in 6 years, is broken by this PR - and something in the rustc workspace depends on it.

Crates.io does not even show separate download stats for this version; "other" is at around 10k downloads on workdays. A lot of that might just be rustc builds thougg. (The latest version is at 100k downloads.)

**Comment by scottmcm** [comment]
Found it: we're apparently building a version of @XAMPPRocky's tokei from Feb, which uses an old winapi
```text
├── term_size v0.3.1
│   ├── kernel32-sys v0.2.2
│   │   └── winapi v0.2.8
│   │   [build-dependencies]
│   │   └── winapi-build v0.1.1
│   └── winapi v0.2.8
```

master has fixed it, though:
```text
├── term_size v0.3.2
│   └── winapi v0.3.9
```

So I think this sha needs to be bumped:

https://github.com/rust-lang/rust/blob/56f132565eb31eeb9ec7e1800a6ab2ca354e710e/src/tools/cargotest/main.rs#L40-L42

**Comment by RalfJung** [comment]
@scottmcm is there a process associated with bumping cargotest crates, or can I just do it in this PR?

**Comment by scottmcm** [comment]
Not a clue!  Never seen that file before in my life; just found it with ripgrep.

It looks like it's just there to smoke that cargo isn't severely broken (<https://rustc-dev-guide.rust-lang.org/tests/intro.html#cargo-test>) so I vote we just bump it and see if anyone yells at me for saying it's fine 🙃

@bors r=cjgillot,scottmcm

**Comment by RalfJung** [comment]
I was going to wait a bit before landing this to give the winapi maintainer time for a patch release, should they choose to make one (https://github.com/retep998/winapi-rs/issues/1030).

@bors r-

**Comment by RalfJung** [comment]
We're branching a new train from master on Oct 28 (according to the forge), so right after that happened would probably be a good time to land this.

**Comment by RalfJung** [comment]
https://github.com/rust-lang/rust/pull/103731 landed so let's go ahead with this one.
@bors r=cjgillot,scottmcm

**Comment by RalfJung** [comment]
Another use of old winapi in [iron](https://github.com/iron/iron).

**Comment by RalfJung** [comment]
Iron does not contain a lockfile, and even on iron master it still has a dependency on winapi 0.2. So either we need a relase of winapi 0.2 that fixes the build or we need to drop iron from cargotest...

Cc @retep998

**Comment by jplatte** [comment]
FWIW, if the upgrade is simple enough iron could be upgraded. I don't think there's a chance it will be *properly* maintained again, but I know sbd. with publish rights who could do a point release for upgraded deps, if somebody wants to invest some time to do basic maintenance like dependency upgrades I'm sure we could work out the release of another version.

**Comment by RalfJung** [comment]
The dependency is going through various indirections... unfortunately `cargo tree` does not seem to work for target-specific dependencies, so I don't have the full picture, but one chain is "hyper 0.12 -> tokio 0.1 -> mio 0.6 -> winapi 0.2". Hyper 0.12 still has [3k downloads on weekdays](https://crates.io/crates/hyper/0.12.36) so breaking it is probably not an option...

Not sure if a mio 0.6 patch release that bumps winapi is realistic. Honestly a winapi 0.2 patch release is probably the least amount of work.

**Comment by RalfJung** [comment]
I wrote a [winapi patch that fixes the build](https://github.com/RalfJung/winapi-rs/commit/7d928c3a8b82954e5f61e0d6b4ab83f2767da9c8) and pushed it to https://github.com/RalfJung/winapi-rs/tree/0.2. If we could just get a winapi 0.2.9 release with that patch, that'd be amazing. :D

**Comment by ChrisDenton** [comment]
```
> cargo tree -i winapi@0.2.8
winapi v0.2.8
├── kernel32-sys v0.2.2
│   ├── mio v0.6.23
│   │   ├── tokio v0.1.22
│   │   │   └── hyper v0.12.36
│   │   │       ├── iron v0.6.0 (R:\Crates\iron\iron)
│   │   │       │   ├── iron-exmaples v0.1.0 (R:\Crates\iron\examples)
│   │   │       │   ├── mount v0.4.0 (R:\Crates\iron\mount)
│   │   │       │   ├── persistent v0.4.0 (R:\Crates\iron\persistent)
│   │   │       │   └── router v0.6.0 (R:\Crates\iron\router)
│   │   │       └── iron-exmaples v0.1.0 (R:\Crates\iron\examples)
│   │   ├── tokio-reactor v0.1.12
│   │   │   ├── hyper v0.12.36 (*)
│   │   │   ├── tokio v0.1.22 (*)
│   │   │   └── tokio-tcp v0.1.4
│   │   │       └── hyper v0.12.36 (*)
│   │   └── tokio-tcp v0.1.4 (*)
│   └── miow v0.2.2
│       └── mio v0.6.23 (*)
├── mio v0.6.23 (*)
├── miow v0.2.2 (*)
└── ws2_32-sys v0.2.1
    └── miow v0.2.2 (*)
```

**Comment by jplatte** [comment]
Yeah but upgrading hyper is very involved and was attempted before. Plus I'm pretty sure upgrading hyper means upgrading tokio to a semver-incompatible version, which is a breaking change.

I would guess mio doing a 0.6.x release to upgrade winapi would be easiest, but tokio doing a 0.1.x release to upgrade mio also doesn't sound entirely unrealistic.

**Comment by RalfJung** [comment]
winapi doing a 0.2.x release still sounds like the best solution to me.^^

**Comment by jplatte** [comment]
Errr yes, I didn't exclude that option intentionally when writing the last comment 😅

Did somebody already ping the maintainer / is there a reason why that branch is not a PR?

**Comment by RalfJung** [comment]
> I would guess mio doing a 0.6.x release to upgrade winapi would be easiest

Not knowing that part of the ecosystem at all, that's not something I will be able to do, but if it's actually easy for someone reading along then that might be an option. mio 0.6.23 still has 15k downloads on weekdays. (winapi 0.2.8 has about 10k.)

**Comment by RalfJung** [comment]
Cc @Thomasdezeeuw @Noah-Kennedy  on the question of whether a release of mio 0.6.x that bumps the winapi dependency to 0.3 would be feasible.

**Comment by Thomasdezeeuw** [comment]
> Cc @Thomasdezeeuw @Noah-Kennedy on the question of whether a release of mio 0.6.x that bumps the winapi dependency to 0.3 would be feasible.

Mio v0.6 is not really supported any more, but we can make an exception here I suppose.

**Comment by Noah-Kennedy** [comment]
I'm with @Thomasdezeeuw here. While we don't support v0.6 anymore, something like this would definitely justify a patch release.

**Comment by Manishearth** [comment]
@bors r-

unclear why bors still thinks this is in queue

**Comment by RalfJung** [comment]
Still hoping that we'll hear from @retep998 (does anyone know other ways to reach them?), but meanwhile -- here is a mio patch: https://github.com/tokio-rs/mio/pull/1629.

**Comment by Nilstrieb** [comment]
Could `#[derive(Debug)]` put a `T: Copy` assertion in the generated code instead of requiring the derive to be present?

**Comment by RalfJung** [comment]
I'm not an expert in our `derive` infrastructure, so -- not sure. It could also lead to surprises when someone writes something like
```rust
#[derive(Debug)]
#[repr(packed)]
struct MyStr(String);
```
and there is no error but the generated `Debug` implementation is just not usable. Though maybe taht can somehow be resolved? After all this also prints a nice error
```rust
struct MyType;

#[derive(Debug)]
struct SndType(MyType);
```

Being more clever wrt. packed in our derive macros would also help with [this problem](https://github.com/rust-lang/rust/issues/82523#issuecomment-1311419639).

**Comment by tbu-** [comment]
> Could `#[derive(Debug)]` put a `T: Copy` assertion in the generated code instead of requiring the derive to be present?

Ah, you already had this idea (https://github.com/retep998/winapi-rs/issues/1030#issuecomment-1364755365).

Perhaps this could be implemented by calling a function which requires the struct to be `Copy`.

```
impl fmt::Debug for UnitStruct {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        fn ensure_copy<T: Copy>(_: &T) {}
        ensure_copy(self);
        // rest of the generated debug impl
    }
}
```

This would fail to compile if `Copy` wasn't implemented for the struct.

**Comment by RalfJung** [comment]
https://github.com/rust-lang/rust/pull/104429 landed. Old winapi no longer be broken by this PR, so we can finally give this another go. :)

@bors r=cjgillot,scottmcm

**Pull request merged**